crypto/asn1/a_object.c

```diff
@@ -18,6 +18,7 @@
 #include "internal/asn1_int.h"
 #include "asn1_locl.h"
 
+#if !defined(FIPS_MODE)
```

Comment:
We _really_ don't want _any_ of the ASN1 code inside the FIPS boundary.  Not even including the internal headers.  Any necessary ASN1 work needs to be done via callbacks or hardwired.  The latter is the usual way.

Reply:
Except that `ASN1_OBJECT` is the fundamental type to store the data associated with the NIDs created with the routines in `crypto/objects/`...  Of course, we could use an entirely different NID ↔ name database within the FIPS module, but that's a bigger effort and double maintenance of similar features.

Reply:
Is a NID↔name database even needed inside the FIPS boundary?  Internally, it will use NIDs won't it?  The name resolution should happen outside of the boundary, it is not a cryptographic operation.

If the database is required, would it be possible to treat `ASN1_OBJECT` as opaque pointer and use callbacks?

I agree that a parallel database isn't the way to go.

Reply:
> The name resolution should happen outside of the boundary, it is not a cryptographic operation.

There will be EVP_XX_fetch() calls from within the boundary. For example in the self tests. This requires name to nid resolution.

> We really don't want any of the ASN1 code inside the FIPS boundary.

For all practical purposes there isn't any. We're not bringing in any code that generates or parses ASN.1. The *only* code we are bringing in here is the ASN1_OBJECT_new() function and the ASN1_OBJECT_free() function. Nothing else. I don't really see why that should be considered a problem.

Reply:
Good point about the fetches.

The new and free aren't the main concern.  The internal ASN1_OBJECT structure is #included which means that entire header is inside the boundary. Splitting the header might help as would creating accessor functions.

FIPS only has a couple of dozen algorithms in total and only a subset of these will be referenced by the self tests, perhaps a parallel name↔NID function is workable (thinking a string of `if strcmp else if strcmp` rather than anything fancy).

Reply:
@slontis do you have the list of power up self tests that are necessary?

Reply:
design doc is the closest we have to a list..
Its not just the self tests though right? its anything that contains multiple layers of algs (like HMAC)

Reply:
Agreed.  I figured the self test list would be a beginning towards what names would need to be mapped.

HMAC doesn't necessarily have to resolve the digest name -- it could take a NID directly.  This will depend on the OSSL_PARAM definition that is finally used -- either NID or name could be passed through.  I prefer not exposing NIDs to users but providers will need to know about them.

Reply:
I suggest that we split the NID handling ASN1 APIs into a new OSSL_NID api set and leave behind wrapper functions.
Disconnect the mapping logic from the ASN1 logic itself - they are independent - so we should make the APIs show it. We should not have any ASN1 headers used in the providers - its too easy to make mistakes like that.

And this does demonstrate that we should split out the NID handling into a separate namespace as such.

It will look a whole lot cleaner if that is done IMHO.

Reply:
And users shouldn't see NIDs in the API at all ... so where do we actually need that logic?
i.e. if we need it then split it out, but why do we need it at all in a provider?

Reply:
It should certainly be possible to sseparate the NID ↔ name map from the NID ↔ OID map.  The ASN1_OBJECT lumps them all together, but that's not really necessary.

Reply:
As a matter of fact, all we need is a number ↔ name map in `crypto/evp/evp_fetch.c`, which doesn't anything fancier than a LHASH indexed by name and with a \< number, name \> tuple as value.  The numbers we use in `evp_fetch.c` can all be created on the fly, it's not like the numbers themselves carry any more meaning than being a current index into a dynamic database.

Reply:
#8878 is my attempt to get away from using the ASN1_OBJECT store.  Note that it's currently a draft, I have seen it build, but have made no test or documentation yet...  To be finished next week if that PR is acceptable.

Reply:
Is it prudent to introduce a _new_ set of integers to map algorithm names too?  That's what NIDs are doing isn't it?

The only advantage, I can see, of having another integer is that they then can't be used as NIDs.  I'm fine with & would encourage this but I don't think it makes sense to complicate the code to get it.

If we insist on on using different numbers, they can be converted algorithmically (e.g. the NID * day-of-month the library was built) rather than having a map.

Reply:
This was the intent from the very start when I made the constructors.  The numeric identities are to be seen as session data rather than constants.  The *only* reason I dealt with numeric identities at all was that the method store uses numeric identities as input rather than names.

Reply:
> I prefer not exposing NIDs to users but providers will need to know about them.

I'm a bit confused about why even providers need to know about NIDs.

> The numeric identities are to be seen as session data rather than constants

E.g., in this mindset we only have ephemeral name/number mappings (which could potentially even be unique per provider, if we want)

Reply:
@kaduk you are right, providers shouldn't need to know about NIDs.